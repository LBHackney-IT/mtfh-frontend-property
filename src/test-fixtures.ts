import { Asset, RentGroup } from "@mtfh/common/lib/api/asset/v1/types";

export const testValidDwellingFixture: Asset = {
  id: "2d13b5cb-baf2-91fd-c231-8c5c2ee9548c",
  assetId: "00023400",
  assetType: "Dwelling",
  rootAsset: "14edf718-19ff-ff43-4679-f8ef404fa029",
  rentGroup: RentGroup.HRA,
  parentAssetIds:
    "656feda1-896f-b136-da84-163ee4f1be6c#14edf718-19ff-ff43-4679-f8ef404fa029#a65918aa-e441-bc34-e49d-fa8671f768da",
  assetLocation: {
    floorNo: "2",
    totalBlockFloors: 10,
    parentAssets: [
      {
        type: "Estate",
        id: "14edf718-19ff-ff43-4679-f8ef404fa029",
        name: "Frampton Park Estate",
      },
      {
        type: "High Rise Block (6 storeys or more)",
        id: "a65918aa-e441-bc34-e49d-fa8671f768da",
        name: "1-93 Pitcairn House",
      },
      {
        type: "NA",
        id: "656feda1-896f-b136-da84-163ee4f1be6c",
        name: "Hackney Homes Limited",
      },
    ],
  },
  assetAddress: {
    uprn: "100023022306",
    postPreamble: "",
    addressLine1: "12 PITCAIRN HOUSE",
    addressLine2: "ST THOMASS SQUARE",
    addressLine3: "HACKNEY",
    addressLine4: "LONDON",
    postCode: "E9 6PT",
  },
  assetManagement: {
    agent: "Hackney Homes",
    areaOfficeName: "Homerton (1) Panel Area Team",
    isCouncilProperty: true,
    managingOrganisation: "London Borough of Hackney",
    managingOrganisationId: "c01e3146-e630-c2cd-e709-18ef57bf3724",
    owner: "LBH",
    isTMOManaged: false,
  },
  assetCharacteristics: {
    numberOfBedrooms: 2,
    numberOfLifts: 1,
    numberOfLivingRooms: 1,
    windowType: "DBL",
    yearConstructed: "2077",
    numberOfSingleBeds: 1,
    numberOfDoubleBeds: 1,
    numberOfFloors: 2,
    totalBlockFloors: 5,
    heating: "FCB",
    propertyFactor: "4.5",
    architecturalType: "PRE45MR-FLT",
  },
  tenure: {
    id: "35ffe2a1-ee7f-c09d-df11-c065d97fa7d0",
    paymentReference: "7647012704",
    type: "Secure",
    startOfTenureDate: "2005-08-29T00:00:00Z",
    endOfTenureDate: "2028-04-03T00:00:00Z",
    isActive: true,
  },
};

export interface ValidChildAsset extends Asset {
  isAssetCautionaryAlerted: boolean;
  isActive: boolean;
  assetStatus: null;
  assetContract: null;

  assetManagement: {
    agent: string;
    areaOfficeName: string;
    isCouncilProperty: boolean;
    managingOrganisation: string;
    managingOrganisationId: string;
    owner: string;
    isTMOManaged: boolean;
    propertyOccupiedStatus: string;
    isNoRepairsMaintenance: boolean;
    isTemporaryAccommodation: boolean;
  };
}

export const testValidChildAssetFixture: ValidChildAsset[] = [
  {
    id: "77ec7c49-1ec5-0ef5-bd51-3fbccce7e96e",
    assetId: "00049158",
    assetType: "LettableNonDwelling",
    isAssetCautionaryAlerted: false,
    rentGroup: RentGroup.HRA,
    assetAddress: {
      uprn: "00049158",
      addressLine1: "Gge 12 Pitcairn House St Thomass Square",
      addressLine2: "Hackney",
      addressLine3: "London",
      addressLine4: "",
      postCode: "E9 6PT",
      postPreamble: "",
    },
    tenure: {
      id: "1d7e6192-5697-51c1-760b-593a9e61303b",
      paymentReference: "3749158504",
      startOfTenureDate: "2020-07-06T00:00:00",
      endOfTenureDate: "2021-07-05T00:00:00",
      type: "Private Garage",
      isActive: true,
    },
    rootAsset: "ROOT",
    isActive: false,
    parentAssetIds: "2d13b5cb-baf2-91fd-c231-8c5c2ee9548c",
    assetCharacteristics: {
      numberOfBedrooms: 2,
      numberOfLifts: 1,
      numberOfLivingRooms: 1,
      windowType: "DBL",
      yearConstructed: "2077",
      numberOfSingleBeds: 1,
      numberOfDoubleBeds: 1,
      numberOfFloors: 2,
      totalBlockFloors: 5,
      heating: "FCB",
      propertyFactor: "4.5",
      architecturalType: "PRE45MR-FLT",
    },
    assetManagement: {
      agent: "Hackney Homes",
      areaOfficeName: "Homerton (1) Panel Area Team",
      isCouncilProperty: true,
      managingOrganisation: "London Borough of Hackney",
      managingOrganisationId: "c01e3146-e630-c2cd-e709-18ef57bf3724",
      owner: "LBH",
      isTMOManaged: false,
      propertyOccupiedStatus: "OC",
      isNoRepairsMaintenance: false,
      isTemporaryAccommodation: false,
    },
    assetStatus: null,
    assetLocation: {
      floorNo: "0",
      totalBlockFloors: 10,
      parentAssets: [],
    },
    assetContract: null,
  },
];
