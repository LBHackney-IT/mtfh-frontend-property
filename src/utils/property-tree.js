import React, { Component } from 'react';
import SortableTree from 'react-sortable-tree';
import FileExplorerTheme from 'react-sortable-tree-theme-file-explorer';
import 'react-sortable-tree/style.css';

export const PropertyTree = (props) => {
  const {asset, childAssets, parentAssets} = props;

  const excludedTreeAssets = "656feda1-896f-b136-da84-163ee4f1be6c"

  console.log(`My details are: ${JSON.stringify(asset)}`)
  console.log(`My children details are: ${JSON.stringify(childAssets)}`)
  console.log(`My parent details are: ${JSON.stringify(asset.assetLocation.parentAssets)}`)

  const childNodes = [];

  // Add children
  if (childAssets) {
    for (const [i, v] of childAssets.entries()) {
      
       if (i < 5) {
        console.log(`Adding regular child prompt: ${JSON.stringify(v.id)}`)
        childNodes.push({ title: v.assetAddress.addressLine1, children: [] })
        continue;
      }
      if (i === 5) {
        console.log(`Adding MORE prompt: ${JSON.stringify(v.id)}`)
        childNodes.push({ title: '<< MORE.... >>', children: [] })
        continue;
      }
      if (i > 5) {
        console.log(`Too many children, not adding prompt: ${JSON.stringify(v.id)}`)
        continue;
      }
    }
  }

  const treeData = [];

  //Generate me
  var principle = { title: 'Principle', children: [{ title: asset.assetAddress.addressLine1, children: [...childNodes], expanded: true}], expanded: true }

  // Add parents and principle
  if (asset.assetLocation.parentAssets) {

    
    let validParents = asset.assetLocation.parentAssets.filter((el) => !excludedTreeAssets.includes(el.id)) 
    
    for (const [i, v] of validParents.entries()) {
      if (i === validParents.length - 1) {
        console.log(`Adding principle: ${JSON.stringify(v.id)}`)
        principle.title = v.name;
        treeData.push(principle)
      }
      else 
      {
        console.log(`Adding parent: ${JSON.stringify(v.id)}`)
        treeData.push({ title: v.name, children: [] })
      }
    }
  }

  const onChangeHander = (e) => {
    console.log(e)
  }
    return (
      <div style={{ height: 800 }}> 
        <SortableTree
          treeData={treeData}
          onChange={onChangeHander}
          //theme={FileExplorerTheme}
        />
      </div>
    ); 
}

const generateNode = (name, childList) => {
  return { title: name, children: [childList]  }
}