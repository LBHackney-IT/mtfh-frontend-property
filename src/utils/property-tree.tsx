import React, { Component } from "react";
// @ts-ignore
import SortableTree from "react-sortable-tree";

import { Asset } from "@mtfh/common/lib/api/asset/v1/types";
import "react-sortable-tree/style.css";

interface PropertyTreeProps {
  asset: Asset;
  childAssets: Asset[] | undefined;
  parentAssets: Asset[] | undefined;
}

export const PropertyTree = (props: PropertyTreeProps): JSX.Element => {
  const { asset, childAssets, parentAssets } = props;

  const excludedTreeAssets = "656feda1-896f-b136-da84-163ee4f1be6c"; // Hackney Homes

  const childNodes = [];

  // Add children
  if (childAssets) {
    for (const [i, v] of childAssets.entries()) {
      if (i < 5) {
        childNodes.push(generateNode(v.assetAddress.addressLine1, [], v.id));
        continue;
      }
      if (i === 5) {
        childNodes.push({ title: "<< MORE.... >>", children: [] });
        continue;
      }
      if (i > 5) {
        continue;
      }
    }
  }

  const treeData = [];

  //Generate principle
  const principle = {
    title: <span>Principle</span>,
    children: [
      {
        title: `${asset.assetAddress.addressLine1} (this asset)`,
        children: [...childNodes],
        expanded: true,
      },
    ],
    expanded: true,
  };

  // Add parents and principle
  if (asset.assetLocation.parentAssets) {
    const validParents = asset.assetLocation.parentAssets.filter(
      (el) => !excludedTreeAssets.includes(el.id),
    );

    for (const [i, v] of validParents.entries()) {
      // Attach princple to last parent
      if (i === validParents.length - 1) {
        principle.title = <a href={`/property/${v.id}`}>{v.name}</a>;
        treeData.push(principle);
      } else {
        treeData.push(generateNode(v.name, [], v.id));
      }
    }
  }

  const onChangeHander = (e: Event) => {
    console.log(e);
  };
  return (
    <div style={{ height: 300 }}>
      <SortableTree treeData={treeData} onChange={onChangeHander} />
    </div>
  );
};

const generateNode = (name: string, childList: Array<JSX.Element>, id: string) => {
  const node = <a href={`/property/${id}`}>{name}</a>;

  return { title: node, children: [childList] };
};
