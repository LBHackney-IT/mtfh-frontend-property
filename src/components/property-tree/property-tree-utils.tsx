import React from "react";

import { ValidChildAsset } from "../../utils/test-fixtures";

import { Asset } from "@mtfh/common/lib/api/asset/v1/types";

interface TreeAsset {
  title: React.ReactNode;
  children: Array<TreeAsset> | null | { title: string; children: null };
  expanded?: boolean;
}

const addChildrenAssets = (
  childAssets: Asset[] | ValidChildAsset[] | undefined,
): TreeAsset[] => {
  const childrenAssets: TreeAsset[] = [];

  if (childAssets) {
    for (const [i, v] of childAssets.entries()) {
      if (i < 5) {
        childrenAssets.push(generateNode(v.assetAddress.addressLine1, [], v.id));
      } else if (i === 5) {
        childrenAssets.push({ title: "<< MORE.... >>", children: null });
      } else {
        break;
      }
    }
  }

  return childrenAssets;
};

const generateNode = (name: string, childList: TreeAsset[], id: string): TreeAsset => {
  const node = (
    <a className="lbh-link govuk-link" href={`/property/${id}`}>
      {name}
    </a>
  );

  return { title: node, children: childList, expanded: false };
};

const generatePrinciple = (asset: Asset, childNodes: TreeAsset[]): TreeAsset => {
  return {
    title: (
      <span>{asset.assetLocation?.parentAssets?.length ? "Principle" : "Hackney"}</span>
    ),
    children: [
      {
        title: `${asset.assetAddress.addressLine1} (this asset)`,
        children: childNodes,
        expanded: true,
      },
    ],
    expanded: true,
  };
};

const addParentsAndPrinciple = (
  asset: Asset,
  childNodes: TreeAsset[],
  excludedTreeAssets: string,
  principle: TreeAsset,
) => {
  const treeViewElements: TreeAsset[] = [];

  if (asset.assetLocation?.parentAssets?.length) {
    const validParents = asset.assetLocation.parentAssets.filter(
      (el) => !excludedTreeAssets.includes(el.id),
    );

    for (const [i, v] of validParents.entries()) {
      // Attach principle to last parent
      if (i === validParents.length - 1) {
        principle.title = (
          <a className="lbh-link govuk-link" href={`/property/${v.id}`}>
            {v.name}
          </a>
        );
        treeViewElements.push(principle);
      } else {
        treeViewElements.push(generateNode(v.name, [], v.id));
      }
    }
  } else {
    treeViewElements.push(generatePrinciple(asset, childNodes));
  }

  return treeViewElements;
};

export { TreeAsset, addChildrenAssets, generatePrinciple, addParentsAndPrinciple };
