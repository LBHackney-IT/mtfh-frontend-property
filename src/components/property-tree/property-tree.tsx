import React, { useEffect, useState } from "react";
// @ts-ignore
import SortableTree from "react-sortable-tree";

import { ValidChildAsset } from "../../utils/test-fixtures";
import {
  TreeAsset,
  addChildrenAssets,
  addParentsAndPrinciple,
  generatePrinciple,
} from "./property-tree-utils";

import { Asset } from "@mtfh/common/lib/api/asset/v1/types";

import "react-sortable-tree/style.css";

interface PropertyTreeProps {
  asset: Asset;
  childAssets: Asset[] | ValidChildAsset[] | undefined;
}

export const PropertyTree = ({ asset, childAssets }: PropertyTreeProps): JSX.Element => {
  const excludedTreeAssets = "656feda1-896f-b136-da84-163ee4f1be6c"; // Hackney Homes

  const [treeViewData, setTreeViewData] = useState<TreeAsset[]>([]);
  const [childNodes, setChildNodes] = useState<TreeAsset[]>([]);

  useEffect(() => {
    setChildNodes(addChildrenAssets(childAssets, asset.id));
  }, [childAssets]);

  useEffect(() => {
    const principle = generatePrinciple(asset, childNodes);
    setTreeViewData(
      addParentsAndPrinciple(asset, childNodes, excludedTreeAssets, principle),
    );
  }, [childNodes, asset]);

  const onChangeHandler = (treeData: TreeAsset[]) => {
    setTreeViewData(treeData);
  };

  return (
    <div>
      <SortableTree
        treeData={treeViewData}
        onChange={onChangeHandler}
        isVirtualized={false}
      />
    </div>
  );
};
