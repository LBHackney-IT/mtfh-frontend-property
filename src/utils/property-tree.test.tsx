import React from "react";

import { mockAssetV1, render, server } from "@hackney/mtfh-test-utils";
import { screen } from "@testing-library/react";
import { rest } from "msw";

import { testValidChildAssetFixture, testValidDwellingFixture } from "../test-fixtures";
import { PropertyTree } from "./property-tree";

beforeEach(() => {
  jest.resetAllMocks();
  server.use(
    rest.get(
      `/api/v1/cautionary-alerts/properties-new/${mockAssetV1.assetId}`,
      (req, res, ctx) => res(ctx.status(200), ctx.json({ alerts: [] })),
    ),
  );
});

test("renders the property tree correctly", async () => {
  const { container } = render(
    <PropertyTree
      asset={testValidDwellingFixture}
      childAssets={testValidChildAssetFixture}
    />,
    {},
  );

  expect(container).toMatchSnapshot();

  screen.getByText("12 PITCAIRN HOUSE (this asset)");
  screen.getByText("Gge 12 Pitcairn House St Thomass Square");
  screen.getByText("1-93 Pitcairn House");
});

test("doesn't render the Hackney Housing ancestor", async () => {
  render(
    <PropertyTree
      asset={testValidDwellingFixture}
      childAssets={testValidChildAssetFixture}
    />,
    {},
  );

  const hhElement = screen.queryByText("Hackney Homes Limited");

  expect(hhElement).toBeNull();
});

test("renders components with links", async () => {
  render(
    <PropertyTree
      asset={testValidDwellingFixture}
      childAssets={testValidChildAssetFixture}
    />,
    {},
  );

  expect(screen.getByText("1-93 Pitcairn House").getAttribute("href")).toBe(
    `/property/a65918aa-e441-bc34-e49d-fa8671f768da`,
  );
});
