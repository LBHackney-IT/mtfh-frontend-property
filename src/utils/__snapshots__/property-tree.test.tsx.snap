// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the property tree correctly 1`] = `
<div>
  <div
    style="height: 300px;"
  >
    <div
      class="rst__tree"
    >
      <div
        class="rst__node"
        style="height: 62px;"
      >
        <div
          class="rst__lineBlock rst__lineHalfHorizontalRight rst__lineHalfVerticalBottom"
          style="width: 44px;"
        />
        <div
          class="rst__nodeContent"
          style="left: 44px;"
        >
          <div
            style="height: 100%;"
          >
            <div>
              <button
                aria-label="Expand"
                class="rst__expandButton"
                style="left: -22px;"
                type="button"
              />
            </div>
            <div
              class="rst__rowWrapper"
            >
              <div
                class="rst__row"
                style="opacity: 1;"
              >
                <div
                  class="rst__moveHandle"
                  draggable="true"
                />
                <div
                  class="rst__rowContents"
                >
                  <div
                    class="rst__rowLabel"
                  >
                    <span
                      class="rst__rowTitle"
                    >
                      <a
                        class="lbh-link govuk-link"
                        href="/property/14edf718-19ff-ff43-4679-f8ef404fa029"
                      >
                        Frampton Park Estate
                      </a>
                    </span>
                  </div>
                  <div
                    class="rst__rowToolbar"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rst__node"
        style="height: 62px;"
      >
        <div
          class="rst__lineBlock rst__lineHalfVerticalTop rst__lineHalfHorizontalRight"
          style="width: 44px;"
        />
        <div
          class="rst__nodeContent"
          style="left: 44px;"
        >
          <div
            style="height: 100%;"
          >
            <div>
              <button
                aria-label="Collapse"
                class="rst__collapseButton"
                style="left: -22px;"
                type="button"
              />
              <div
                class="rst__lineChildren"
                style="width: 44px;"
              />
            </div>
            <div
              class="rst__rowWrapper"
            >
              <div
                class="rst__row"
                style="opacity: 1;"
              >
                <div
                  class="rst__moveHandle"
                  draggable="true"
                />
                <div
                  class="rst__rowContents"
                >
                  <div
                    class="rst__rowLabel"
                  >
                    <span
                      class="rst__rowTitle"
                    >
                      <a
                        class="lbh-link govuk-link"
                        href="/property/a65918aa-e441-bc34-e49d-fa8671f768da"
                      >
                        1-93 Pitcairn House
                      </a>
                    </span>
                  </div>
                  <div
                    class="rst__rowToolbar"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rst__node"
        style="height: 62px;"
      >
        <div
          class="rst__lineBlock"
          style="width: 44px;"
        />
        <div
          class="rst__lineBlock rst__lineHalfVerticalTop rst__lineHalfHorizontalRight"
          style="width: 44px;"
        />
        <div
          class="rst__nodeContent"
          style="left: 88px;"
        >
          <div
            style="height: 100%;"
          >
            <div>
              <button
                aria-label="Collapse"
                class="rst__collapseButton"
                style="left: -22px;"
                type="button"
              />
              <div
                class="rst__lineChildren"
                style="width: 44px;"
              />
            </div>
            <div
              class="rst__rowWrapper"
            >
              <div
                class="rst__row"
                style="opacity: 1;"
              >
                <div
                  class="rst__moveHandle"
                  draggable="true"
                />
                <div
                  class="rst__rowContents"
                >
                  <div
                    class="rst__rowLabel"
                  >
                    <span
                      class="rst__rowTitle"
                    >
                      12 PITCAIRN HOUSE (this asset)
                    </span>
                  </div>
                  <div
                    class="rst__rowToolbar"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rst__node"
        style="height: 62px;"
      >
        <div
          class="rst__lineBlock"
          style="width: 44px;"
        />
        <div
          class="rst__lineBlock"
          style="width: 44px;"
        />
        <div
          class="rst__lineBlock rst__lineHalfVerticalTop rst__lineHalfHorizontalRight"
          style="width: 44px;"
        />
        <div
          class="rst__nodeContent"
          style="left: 132px;"
        >
          <div
            style="height: 100%;"
          >
            <div>
              <button
                aria-label="Expand"
                class="rst__expandButton"
                style="left: -22px;"
                type="button"
              />
            </div>
            <div
              class="rst__rowWrapper"
            >
              <div
                class="rst__row"
                style="opacity: 1;"
              >
                <div
                  class="rst__moveHandle"
                  draggable="true"
                />
                <div
                  class="rst__rowContents"
                >
                  <div
                    class="rst__rowLabel"
                  >
                    <span
                      class="rst__rowTitle"
                    >
                      <a
                        class="lbh-link govuk-link"
                        href="/property/77ec7c49-1ec5-0ef5-bd51-3fbccce7e96e"
                      >
                        Gge 12 Pitcairn House St Thomass Square
                      </a>
                    </span>
                  </div>
                  <div
                    class="rst__rowToolbar"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
